src/Opaleye/With.hs

```diff
@@ -0,0 +1,108 @@
+{-# LANGUAGE FlexibleContexts #-}
+
+module Opaleye.With
+  ( with
+  , withExplicit
+  , withRecursive
+  , withRecursiveExplicit
+  )
+where
+
+import Data.List.NonEmpty (NonEmpty((:|)))
+import Data.Profunctor.Product.Default (Default, def)
+import Opaleye.Internal.Binary (Binaryspec(..), runBinaryspec)
+import Opaleye.Internal.QueryArr (Select, runSimpleQueryArr', productQueryArr')
+import Opaleye.Internal.Unpackspec (Unpackspec(..), runUnpackspec)
+import Opaleye.Internal.PackMap (PackMap(..))
+import qualified Opaleye.Internal.HaskellDB.PrimQuery as HPQ
+import qualified Opaleye.Internal.PrimQuery           as PQ
+import qualified Opaleye.Internal.Sql                 as Sql
+import qualified Opaleye.Internal.Tag                 as Tag
+import qualified Opaleye.Internal.PackMap as PM
+
+
+with :: Default Unpackspec a a => Select a -> (Select a -> Select b) -> Select b
+with = withExplicit def
+
+
+withRecursive :: Default Binaryspec a a => Select a -> (a -> Select a) -> Select a
+withRecursive = withRecursiveExplicit def
+
+
+withExplicit :: Unpackspec a a -> Select a -> (Select a -> Select b) -> Select b
+withExplicit unpackspec query f = productQueryArr' $ \_ -> do
+  tableName <- HPQ.Symbol "cte" <$> Tag.fresh
+
+  let
+    mkSelect =
+      PQ.BaseTable (PQ.TableIdentifier Nothing (Sql.sqlSymbol tableName))
+
+  (a, primQ) <- runSimpleQueryArr' query ()
+  startTag <- Tag.fresh
+
+  let
+    (a', bindings) = PM.run $
+      runUnpackspec unpackspec (PM.extractAttr "column" startTag) a
+    primQ' = mkSelect bindings
+    query' = productQueryArr' $ \_ -> pure (a', primQ')
+
+  (b, primQ'') <- runSimpleQueryArr' (f query') ()
+
+  let
+    query'' = PQ.With PQ.NotRecursive tableName primQ primQ''
+
+  pure (b, query'')
+
+
+withRecursiveExplicit :: Binaryspec a a -> Select a -> (a -> Select a) -> Select a
+withRecursiveExplicit binaryspec base recursive = productQueryArr' $ \_ -> do
+  tableName <- HPQ.Symbol "cte" <$> Tag.fresh
+
+  let
+    mkSelect = PQ.BaseTable (PQ.TableIdentifier Nothing (Sql.sqlSymbol tableName))
+
+  (a, primQBase) <- runSimpleQueryArr' base ()
+
+  startTag <- Tag.fresh
+
+  let
+    (columns, _) = PM.run $
+      runUnpackspec unpackspec (PM.extractAttr "binary" startTag) a
```

Comment:
```suggestion
      runUnpackspec unpackspec (PM.extractAttr "base" startTag) a
```

Perhaps a better name?

Reply:
I'm pretty sure it needs to be `binary` because the column names from [here](https://github.com/circuithub/haskell-opaleye/blob/214e2841f03a6ecf3d64138985fb3dc6df787488/src/Opaleye/With.hs#L75) need to match the column names from [here](https://github.com/circuithub/haskell-opaleye/blob/214e2841f03a6ecf3d64138985fb3dc6df787488/src/Opaleye/With.hs#L90).

Reply:
Hmm, there's something funny going on because `extractAttr` ignores the `PrimExpr`s inside the `a` argument to `runUnpackspec`. That suggests that if you delete

```haskell
  let
    (columns, _) = PM.run $
      runUnpackspec unpackspec (PM.extractAttr "binary" startTag) a
```

and use `a` instead of `columns` where the latter appears then you should get the same result. I'll dig into this more when I get some time (hopefully at the weekend) unless you want to investigate before then.

Reply:
If tried it, and it's not possible to use just `columns = a`. I guess because of the recursive nature of the thing we have to get `columns` before the thing that morally produces columns is run. I don't really quite grasp what's going on here, but I'm pretty confident that it works fine, so hopefully I'll be able to merge soon.